packages/excalidraw/change.ts

```diff
@@ -0,0 +1,1343 @@
+      keys = Object.keys(object2);
+    } else {
+      keys = Array.from(
+        new Set([...Object.keys(object1), ...Object.keys(object2)]),
+      );
+    }
+
+    for (const key of keys) {
+      const object1Value = object1[key as keyof T];
+      const object2Value = object2[key as keyof T];
+
+      if (object1Value !== object2Value) {
+        if (
+          !skipShallowCompare &&
+          typeof object1Value === "object" &&
+          typeof object2Value === "object" &&
+          object1Value !== null &&
+          object2Value !== null &&
+          isShallowEqual(object1Value, object2Value)
+        ) {
+          continue;
+        }
+
+        yield key;
+      }
+    }
+  }
+}
+
+/**
+ * Encapsulates the modifications captured as `Delta`/s.
+ */
+interface Change<T> {
+  /**
+   * Inverses the `Delta`s inside while creating a new `Change`.
+   */
+  inverse(): Change<T>;
+
+  /**
+   * Applies the `Change` to the previous object.
+   *
+   * @returns a tuple of the next object `T` with applied change, and `boolean`, indicating whether the applied change resulted in a visible change.
+   */
+  applyTo(previous: Readonly<T>, ...options: unknown[]): [T, boolean];
+
+  /**
+   * Checks whether there are actually `Delta`s.
+   */
+  isEmpty(): boolean;
+}
+
+export class AppStateChange implements Change<AppState> {
+  private constructor(private readonly delta: Delta<ObservedAppState>) {}
+
+  public static calculate<T extends ObservedAppState>(
+    prevAppState: T,
+    nextAppState: T,
+  ): AppStateChange {
+    const delta = Delta.calculate(
+      prevAppState,
+      nextAppState,
+      undefined,
+      AppStateChange.postProcess,
+    );
+
+    return new AppStateChange(delta);
+  }
+
+  public static empty() {
+    return new AppStateChange(Delta.create({}, {}));
+  }
+
+  public inverse(): AppStateChange {
+    const inversedDelta = Delta.create(this.delta.inserted, this.delta.deleted);
+    return new AppStateChange(inversedDelta);
+  }
+
+  public applyTo(
+    appState: Readonly<AppState>,
+    elements: SceneElementsMap,
+  ): [AppState, boolean] {
+    const {
+      selectedElementIds: removedSelectedElementIds = {},
+      selectedGroupIds: removedSelectedGroupIds = {},
+    } = this.delta.deleted;
+
+    const {
+      selectedElementIds: addedSelectedElementIds = {},
+      selectedGroupIds: addedSelectedGroupIds = {},
+      selectedLinearElementId,
+      editingLinearElementId,
+      ...directlyApplicablePartial
+    } = this.delta.inserted;
+
+    const mergedSelectedElementIds = Delta.merge(
+      appState.selectedElementIds,
+      addedSelectedElementIds,
+      removedSelectedElementIds,
+    );
+
+    const mergedSelectedGroupIds = Delta.merge(
+      appState.selectedGroupIds,
+      addedSelectedGroupIds,
+      removedSelectedGroupIds,
+    );
+
+    const selectedLinearElement =
+      selectedLinearElementId && elements.has(selectedLinearElementId)
+        ? new LinearElementEditor(
```

Comment:
Double-check if it makes sense keeping whole `LinearElementEditor instance state as part the `appState`, as from the history point of view it does not.

Reply:
Added TODO